Simplify da layer

MartinOndejka opened the pull request:
What's missing:
- [x] integrate to sequencer
- [ ] add DoS protection #191 
- [ ] add outer account check for da attestation #192

# Node Description

The data availability multisig, as explained here #12, is a very simple idea and previous implementation was extremely complex. We don't need EVM nor the blockchain for executing the simple multisig architecture. In the new simplified solution of DA committee, each node is a standalone service, which on-request stores and signs the batch. The signature is an attestation that the node knows and has the information on how to reproduce the ledger.

The interface of the node API is roughly the following (`t` is the  DA data unit):
```ocaml
type t =
      { source_ledger_hash : Ledger_hash.t
      ; target_ledger_hash : Ledger_hash.t
      ; diff : (int * Account.t) list
      ; command_with_action_step_flags :
          (User_command.t * bool list) option
      }

val post_batch : ledger_openings:Sparse_ledger.t -> batch:t-> Signature.t
val get_batch : Ledger_hash.t -> t option
```

Each data batch has the all the account changes stored in the `diff`. The da node applies each account's diff to the ledger openings, and compares the resulting ledger hash with the `target_ledger_hash`. If the adversary would want to lie, and therefore convince node to sign the ledger hash that the node doesn't know how to construct, the diff would need to be either incorrect, or some would be missing. To ensure neither happens following steps are taken:
1. Check that `root ledger_openings = batch.source_ledger_hash`.
2. Check that `batch.source_ledger_hash` is either in the database or an empty ledger.
3. Check that the indices in `batch.diff` are unique. 
4. Set each account in `batch.diff` to the `ledger_openings` and check that result equals [batch.target_ledger_hash].
5. Sign [batch.target_ledger_hash].
6. Store the batch under the [batch.target_ledger_hash].

For convenience, we store also the full command, which helps to not only reconstruct the ledger, but also have to full history of the state. To ensure that the command is a correct one we do one additional step:
- Check that after applying all the receipts of the command, the receipt chain hashes match the target ledger.

The receipt chain hash is a commitment to all the transaction commitments. Transaction commitment doesn't include authorization fields, so the adversary can lie about the authorization of the corresponding command, this however doesn't allow to convince the da node to sign the incorrect ledger hash.

The api is exposed as `Async.Rpc` which is a typesafe ocaml rpc with binprot serialization. It's supposed to be very fast and reliable.

# Client description

The client of the da node in our case is the sequencer. The sequencer posts each transaction to the every da node and at the event of the state commitment to the L1 takes the signatures of the latest ledger hash as the DA attestation. There's one requirement for the sequencer to meet, the transactions need to come in a strict order. If the transactions with order A, B would come in reversed order, the node would refuse the second transaction as it would not be able to reconstruct the previous state. To do that we use `Da_layer.Client.Sequencer` module, which enqueues all the transactions and makes sure that they are posted in correct order. In the event of commitment the sequencer asks the particular signature and `Da_layer.Client.Sequencer` returns a `Deferred.t` which gets filled once the signatures are present.

MartinOndejka committed:
Delete old da layer

MartinOndejka committed:
Move kvdb to separate library

`module Kvdb` was part of the `sequencer_lib`
and I want to use the same interface for the da_node's db

MartinOndejka committed:
Add new da layer

MartinOndejka committed:
Add syncing of the da node

L-as commented on src/app/zeko/da_layer/cli.ml:
```diff
@@ -0,0 +1,37 @@
+         flag "--node-to-sync" (optional string) ~doc:"string Node to sync with"
+       in
+       fun () ->
+         let signer = Sys.getenv_exn "MINA_PRIVATE_KEY" in
```
This is quite suboptimal FWIW. The private key should not be in the memory of the DA node.
It should make a request to some external service to do the signing.
The external service could then either do it in hardware (e.g. FIDO2, TPM, hardware wallet),
or do it _properly_, that is, making sure that even if the private key is in memory, that it's never copied around, that it's not swapped to disk, etc. There are already many implementations of things in this direction.

Of course, it's complicated by the fact that we use Schnorr on Pallas,
but it should be a minor complication.
At the very least it should be a separate process.

You have to consider the attack vector where someone uses this API
to hack the DA node and get hold of their private key (or use it at least).

L-as replied:
Considering our constraints,
we should use the TPM functionality that the kernel provides.

This however constrains us to non-native curves, so we need to verify non-native signatures.
This is probably worth it however.

MartinOndejka replied:
#189

MartinOndejka committed:
Refactor kvdb with GADT so that key value types need to match

Before, `module Kvdb` was an interface for using correct _key_ type.
Now, the module ensures correct match between _key_ and _value_ types.

Having let's say 2 types stored in kvdb, `Post.t` and `User.t`, with corresponding
key types `Post_id.t` and `User_id.t`, it's not possible to use
`Post_id.t` key with `User.t` value and vice versa.

L-as commented:
Please rebase and force-push

MartinOndejka committed:
Integrate into sequencer

`Batch.t` gained few fields and mainly, the idea changed from having one da "unit" per batch,
to rather have one da "unit" per transaction. The `command_with_action_step_flags` is a command
together with flags, wether the `action_state` was stepped in corresponding account update.

MartinOndejka committed:
Add test for bootstrap

MartinOndejka committed:
Refactor da_node sync

MartinOndejka committed:
Fix readme

MartinOndejka committed:
Fix dummy actions step env

For now, the action step flags are all true, as sequencer currently always do the action step if possible.

MartinOndejka committed:
Add rpc to query the da node signer public key

This will be needed for tests at least, where you want in the test to query the key and deploy
the outer account with correctly set da committee. Might be useful for other things as well, so I've added it already.

MartinOndejka committed:
Add comments

MartinOndejka committed:
Fix rebase

L-as commented:
Do the commits mean anything? Should I look at it individually or the big picture? Also, please add a big PR description. Phrase it in your own terms your understanding of the system. Write docs so I know we're on the same page.

MartinOndejka commented:
> Do the commits mean anything? Should I look at it individually or the big picture? Also, please add a big PR description. Phrase it in your own terms your understanding of the system. Write docs so I know we're on the same page.

It's probably easier to review while following the commits. I've edited commit messages with notes, to understand particular changes in design.

MartinOndejka committed:
Add option to deploy outer account with initial db

L-as commented on src/app/zeko/da_layer/lib/batch.ml:
```diff
@@ -0,0 +1,19 @@
+  module V1 = struct
+    type t =
+      { source_ledger_hash : Ledger_hash.Stable.V1.t
+      ; diff : (int * Account.Stable.V2.t) list
```
Should be uint

MartinOndejka replied:
The ledger uses int, so then there are bunch of annoying conversions.

L-as replied:
Makes sense

L-as commented on src/app/zeko/da_layer/cli.ml:
```diff
@@ -11,14 +12,23 @@ let run_node =
            (optional_with_default 8080 int)
            ~doc:"int Port to listen on"
+       and node_to_sync =
+         flag "--node-to-sync" (optional string) ~doc:"string Node to sync with"
```
I don’t get this, elaborate?

MartinOndejka replied:
If you want to start the node that joins the da layer later, you provide a node uri from which you can sync. Currently it fetches all the data that it's missing.

L-as replied:
There should be several here. Should also write _DA_ node.

MartinOndejka replied:
several?

MartinOndejka replied:
So list of da nodes to sync?

L-as replied:
Yes

L-as commented on src/app/zeko/da_layer/lib/client.ml:
```diff
@@ -30,9 +32,137 @@ let dispatch ?(max_tries = 5) ~logger
+    ; q : unit Async.Sequencer.t
+    ; mutable signatures : Signature.t list Deferred.t Ledger_hash.Map.t
+    ; mutable last_distributed_batch : Ledger_hash.t option
+    }
```
Documentation for all the fields

L-as commented on src/app/zeko/da_layer/lib/client.ml:
```diff
@@ -30,9 +32,137 @@ let dispatch ?(max_tries = 5) ~logger
+    ; last_distributed_batch = None
+    }
+
+  let distribute_batch t ~ledger_openings ~batch =
```
What does this do?

MartinOndejka replied:
I've changed it to `enqueue_distribute_batch` in later commit. `distribute_batch` sends the batch to all the da nodes in the config. Enqueueing is to ensure strict order.

L-as commented on src/app/zeko/da_layer/lib/client.ml:
```diff
@@ -30,9 +32,137 @@ let dispatch ?(max_tries = 5) ~logger
+  in
+  try_first ~accum_errors:[] (Config.nodes config)
+
+let get_ledger_hashes_chain ~logger ~config ~depth ~target_ledger_hash =
```
What does this do

L-as replied:
Comment makes sense

L-as commented on src/app/zeko/da_layer/lib/node.ml:
```diff
@@ -106,32 +128,129 @@ let post_batch t ~ledger_openings ~batch =
+          List.fold_result (Zkapp_command.all_account_updates_list command)
+            ~init:(Unsigned.UInt32.zero, Account_id.Map.empty)
+            ~f:(fun (index, acc) account_update ->
+              (* Receipt chain hash is updated only for account updates authorised with Proof of Signature *)
```
I didn’t know this was the case lol

MartinOndejka replied:
It makes sense that you can't mess with the receipt chain hash of someone else's account.

L-as commented on src/app/zeko/da_layer/cli.ml:
```diff
@@ -16,7 +16,9 @@ let run_node =
-         flag "--testing-mode" no_arg ~doc:" Run in testing mode"
+         flag "--testing-mode" no_arg
+           ~doc:
+             " Run in testing mode, the signer key will be generated randomly"
```
Makes sense

L-as replied:
Rename to —random-sk ?

L-as reviewed (changes requested):
Looks good mostly

L-as commented on src/app/zeko/da_layer/lib/batch.ml:
```diff
@@ -0,0 +1,35 @@
+  module V1 = struct
+    type t =
+      { source_ledger_hash : Ledger_hash.Stable.V1.t
+      ; target_ledger_hash : Ledger_hash.Stable.V1.t
```
Why include target here

MartinOndejka replied:
Hmm yeah this is redundant. I've added it for convenience because sequencer needs to know target ledger hash beforehand to know under which key store the deferred signatures. This however doesn't make much sense.

I'll change the api so that `enqueue_distribute_batch` takes the `target_ledger_hash`.

L-as commented on src/app/zeko/da_layer/lib/batch.ml:
```diff
@@ -0,0 +1,35 @@
+      ; diff : (int * Account.Stable.V2.t) list
+      ; command_with_action_step_flags :
+          (User_command.Stable.V2.t * bool list) option
+      }
```
Comments on all fields in all types as always

L-as commented on src/app/zeko/da_layer/lib/client.ml:
```diff
@@ -0,0 +1,178 @@
+    }
+end
+
+let distribute_batch ~logger ~config ~ledger_openings ~batch ~quorum =
```
Distribute to whom?

MartinOndejka replied:
To all the da nodes.

L-as replied:
Comments please

L-as commented:
I again feel the need for long form documentation in the source code.

MartinOndejka committed:
Rename --testing-mode to --random-sk

MartinOndejka committed:
Remove  from da batch as it's redundant

MartinOndejka committed:
Add readme to the da layer

MartinOndejka committed:
Rename --node-to-sync to --da-node-to-sync

MartinOndejka committed:
Separate client rpc functions

MartinOndejka committed:
Add more comments

MartinOndejka committed:
Rename da batch to diff

MartinOndejka commented:
@L-as I've renamed `batch` as da unit to the `diff`. It was confusing as we are posting to da layer after each transaction and not after every batch.

MartinOndejka committed:
Fix typo

MartinOndejka committed:
Add possibility to sync with the list of da nodes

L-as commented on src/app/zeko/da_layer/README.md:
```diff
@@ -0,0 +1,53 @@
+
+- Check that after applying all the receipts of the command, the receipt chain hashes match the target ledger.
+
+The receipt chain hash is a commitment to all the transaction commitments. Transaction commitment doesn't include authorization fields, so the adversary can lie about the authorization of the corresponding command, this however doesn't allow to convince the da node to sign the incorrect ledger hash.
```
But account update includes the information?

L-as replied:
What is meant here: DA doesn’t know if authorization was provided, but this isn’t relevant to DA.

L-as commented on src/app/zeko/da_layer/lib/batch.ml:
```diff
@@ -7,9 +7,12 @@ module Stable = struct
+            (** List of changed accounts with corresponding index in the ledger *)
       ; command_with_action_step_flags :
           (User_command.Stable.V2.t * bool list) option
+            (** Optionally add command with corresponding action steps to store the history *)
```
Are there not cases where we might want to do multiple commands? Maybe not important though

MartinOndejka replied:
I can't think of any, it can be added in future since the whole api is versioned

L-as commented:
We really need to get some property tests. Probably not now though.
